doc/source/notebooks/advanced/varying_noise.pct.py

```diff
@@ -88,28 +88,38 @@ def generate_data(N=80):
 
 # %%
 class HeteroskedasticGaussian(gpflow.likelihoods.Likelihood):
-    def log_prob(self, F, Y):
+    def __init__(self, **kwargs):
+        # this likelihood expects a single latent function F, and two columns in the data matrix Y:
+        super().__init__(latent_dim=1, observation_dim=2, **kwargs)
+
+    def _log_prob(self, F, Y):
         # log_prob is used by the quadrature fallback of variational_expectations and predict_density.
         # Because variational_expectations is implemented analytically below, this is not actually needed,
         # but is included for pedagogical purposes.
         # Note that currently relying on the quadrature would fail due to https://github.com/GPflow/GPflow/issues/966
-        Y, NoiseVar = Y[:, 0:1], Y[:, 1:2]
+        Y, NoiseVar = Y[:, 0], Y[:, 1]
         return gpflow.logdensities.gaussian(Y, F, NoiseVar)
 
-    def conditional_mean(self, F):
-        raise NotImplementedError
-
-    def conditional_variance(self, F):
-        raise NotImplementedError
-
-    def variational_expectations(self, Fmu, Fvar, Y):
-        Y, NoiseVar = Y[:, 0:1], Y[:, 1:2]
+    def _variational_expectations(self, Fmu, Fvar, Y):
+        Y, NoiseVar = Y[:, 0], Y[:, 1]
         return (
             -0.5 * np.log(2 * np.pi)
             - 0.5 * tf.math.log(NoiseVar)
             - 0.5 * (tf.math.square(Y - Fmu) + Fvar) / NoiseVar
         )
 
+    def _predict_log_density(self, Fmu, Fvar, Y):
+        raise NotImplementedError
+
+    def _predict_mean_and_var(self, Fmu, Fvar):
+        raise NotImplementedError
+
+    def _conditional_mean(self, F):
+        raise NotImplementedError
+
+    def _conditional_variance(self, F):
+        raise NotImplementedError
+
```

Comment:
For the notebook, it might be helpful to add a comment saying that, in general, you would want to implement these but here for demonstration purposes we only implement `_variational_expectations` as that one is the one used for training. The others are only required for prediction.

Reply:
It was decided we wanted to make methods abstract, that's a general design choice for gpflow, I don't think we have to justify everywhere that we do need implementations for abstract classes.
It is like adding an annoying thing and apologising for it

Reply:
We don't have to justify it everywhere, I just think it's helpful to explain in the documentation, and this notebook is the main documentation for "how to write your own likelihood", so I think it'd be good to add a comment - doesn't have to be long.

Reply:
On second thought, I'm wondering if this is just a sign of having overdone it with making them abstractmethods. I only need to implement _conditional_mean and _conditional_variance if I *don't* overwrite predict_mean_and_var directly. What do you think about simply removing the abstractmethod in front of those two, removing their dummy implementations here, and leaving it at that?